### PR TITLE
优化交互，求2.6，2.7测试

### PR DESCRIPTION
**Pull request by leon332157**
新功能我再研究一下

### CODE REVIEW

**Comment by leon332157** [comment]
@chfw @chenjiandongx

File: pyecharts_snapshot/main.py [hunk 1]
```diff
@@ -2,9 +2,9 @@
 import os
 import sys
 import base64
-import subprocess
 from PIL import Image
-
+import subprocess
```
Comment (chenjiandongx) [inline pyecharts_snapshot/main.py]: 为什么要移动下来

File: pyecharts_snapshot/main.py [hunk 2]
```diff
@@ -68,17 +74,15 @@ def make_a_snapshot(file_name, output_name, delay=DEFAULT_DELAY):
         content = io.TextIOWrapper(proc.stdout, encoding="utf-8").read()
     content_array = content.split(',')
     if len(content_array) != 2:
-        raise Exception("No snapshot taken by phantomjs. " +
-                        "Please make sure it is installed " +
-                        "and available on your path")
+        raise OSError("No snapshot taken by phantomjs. Please make sure it is installed and available on your PATH!")
```
Comment (chenjiandongx) [inline pyecharts_snapshot/main.py]: 故意遵循 PEP8 规范而你又给拉长了

File: setup.py [hunk 3]
```diff
@@ -18,7 +17,7 @@
     ]
 }
 DESCRIPTION = (
-    'renders pyecharts as image' +
+    'save pyecharts output file as image' +
```
Comment (chenjiandongx) [inline setup.py]: ？？？

File: pyecharts_snapshot/main.py [hunk 4]
```diff
@@ -105,14 +110,12 @@ def save_as(imagedata, output_name, file_type):
     color = (255, 255, 255)
     b = Image.new('RGB', m.size, color)
     b.paste(m, mask=m.split()[3])
-    b.save(output_name, file_type.upper(), quality=100)
-
+    b.save(output_name, file_type, quality=100)
```
Comment (chenjiandongx) [inline pyecharts_snapshot/main.py]: ？？？

**Comment by leon332157** [comment]
@chenjiandongx 我pycharm自动PEP8

**Comment by leon332157** [comment]
要不我重新排版一下？

**Comment by leon332157** [comment]
@chenjiandongx webbrowser是自带库

**Comment by leon332157** [comment]
@chenjiandongx 你可以试试安装我这个版本，我相信别人用的时候会感觉好些吧

**Comment by leon332157** [comment]
@chenjiandongx 你可以让 @chfw 过一下

**Comment by leon332157** [comment]
import io
import os
import sys
import base64
from PIL import Image
import subprocess
import webbrowser

PY2 = sys.version_info[0] == 2

if PY2:
    from StringIO import StringIO as BytesIO
else:
    from io import BytesIO

PHANTOMJS_EXEC = "phantomjs"
NOT_SUPPORTED_FILE_TYPE = "Do not support file type {}"
DEFAULT_DELAY = 1.5


def main():
    if len(sys.argv) < 2 or len(sys.argv) > 4:
        if len(sys.argv) == 2:
            webbrowser.open('https://github.com/pyecharts/pyecharts-snapshot')
            exit(0)
        else:
            print('''Usage:   snapshot {input file} {output file [png|jpeg|gif|pdf]} {delay_in_seconds}''')
            print('''         snapshot --online_help for help online.''')
            exit(-1)
    try:
        PHANTOMJS_VERSION = (subprocess.check_output([PHANTOMJS_EXEC, '--version'])).decode('utf8')
        print("\nphantomjs version: {}".format(PHANTOMJS_VERSION))
    except OSError:
        print("No phantomjs found in your PATH. Please install it!")
        sys.exit(-1)
    file_name = sys.argv[1]
    delay = DEFAULT_DELAY
    output = 'output.png'
    if len(sys.argv) >= 3:
        file_type = sys.argv[2]
        if file_type in ['pdf', 'jpeg', 'gif']:
            output = 'output.{}'.format(file_type)
        elif file_type != 'png':
            raise IOError(NOT_SUPPORTED_FILE_TYPE.format(file_type))
        if len(sys.argv) == 4:
            delay = float(sys.argv[3])  # in seconds
    make_a_snapshot(file_name, output, delay=delay)


def make_a_snapshot(file_name, output_name, delay=DEFAULT_DELAY):
    print('Generating file ...')
    file_type = output_name.split('.')[-1]
    pixel_ratio = 2
    shell_flag = False
    if sys.platform == 'win32':
        shell_flag = True
    __actual_delay_in_ms = int(delay * 1000)

    # add shell=True and it works on Windows now.
    proc_params = [
        PHANTOMJS_EXEC,
        os.path.join(get_resource_dir('phantomjs'), 'snapshot.js'),
        file_name.replace('\\', '/'),
        file_type,
        str(__actual_delay_in_ms),
        str(pixel_ratio)
    ]
    proc = subprocess.Popen(
        proc_params, stdout=subprocess.PIPE, shell=shell_flag)
    if PY2:
        content = proc.stdout.read()
        content = content.decode('utf-8')
    else:
        content = io.TextIOWrapper(proc.stdout, encoding="utf-8").read()
    content_array = content.split(',')
    if len(content_array) != 2:
        raise OSError("No snapshot taken by phantomjs. Please make sure it is installed and available on your PATH!")
    base64_imagedata = content_array[1]
    imagedata = decode_base64(base64_imagedata.encode('utf-8'))
    if file_type in ['pdf', 'gif']:
        save_as(imagedata, output_name, file_type)
    elif file_type in ['png', 'jpeg']:
        save_as_png(imagedata, output_name)
    else:
        raise IOError(NOT_SUPPORTED_FILE_TYPE.format(file_type))


def decode_base64(data):
    """Decode base64, padding being optional.

    :param data: Base64 data as an ASCII byte string
    :returns: The decoded byte string.

    """
    missing_padding = len(data) % 4
    if missing_padding != 0:
        data += b'=' * (4 - missing_padding)
    return base64.decodestring(data)


def save_as_png(imagedata, output_name):
    with open(output_name, "wb") as f:
        f.write(imagedata)
    print('File saved in {}/{}'.format(os.getcwd(), output_name))


def save_as(imagedata, output_name, file_type):
    m = Image.open(BytesIO(imagedata))
    m.load()
    color = (255, 255, 255)
    b = Image.new('RGB', m.size, color)
    b.paste(m, mask=m.split()[3])
    b.save(output_name, file_type, quality=100)
    print('File saved in {}/{}'.format(os.getcwd(), output_name))


def get_resource_dir(folder):
    current_path = os.path.dirname(__file__)
    resource_path = os.path.join(current_path, folder)
    return resource_path

**Comment by leon332157** [comment]
'''import io
import os
import sys
import base64
from PIL import Image
import subprocess
import webbrowser

PY2 = sys.version_info[0] == 2

if PY2:
    from StringIO import StringIO as BytesIO
else:
    from io import BytesIO

PHANTOMJS_EXEC = "phantomjs"
NOT_SUPPORTED_FILE_TYPE = "Do not support file type {}"
DEFAULT_DELAY = 1.5


def main():
    if len(sys.argv) < 2 or len(sys.argv) > 4:
        if len(sys.argv) == 2:
            webbrowser.open('https://github.com/pyecharts/pyecharts-snapshot')
            exit(0)
        else:
            print('''Usage:   snapshot {input file} {output file [png|jpeg|gif|pdf]} {delay_in_seconds}''')
            print('''         snapshot --online_help for help online.''')
            exit(-1)
    try:
        PHANTOMJS_VERSION = (subprocess.check_output([PHANTOMJS_EXEC, '--version'])).decode('utf8')
        print("\nphantomjs version: {}".format(PHANTOMJS_VERSION))
    except OSError:
        print("No phantomjs found in your PATH. Please install it!")
        sys.exit(-1)
    file_name = sys.argv[1]
    delay = DEFAULT_DELAY
    output = 'output.png'
    if len(sys.argv) >= 3:
        file_type = sys.argv[2]
        if file_type in ['pdf', 'jpeg', 'gif']:
            output = 'output.{}'.format(file_type)
        elif file_type != 'png':
            raise IOError(NOT_SUPPORTED_FILE_TYPE.format(file_type))
        if len(sys.argv) == 4:
            delay = float(sys.argv[3])  # in seconds
    make_a_snapshot(file_name, output, delay=delay)


def make_a_snapshot(file_name, output_name, delay=DEFAULT_DELAY):
    print('Generating file ...')
    file_type = output_name.split('.')[-1]
    pixel_ratio = 2
    shell_flag = False
    if sys.platform == 'win32':
        shell_flag = True
    __actual_delay_in_ms = int(delay * 1000)

    # add shell=True and it works on Windows now.
    proc_params = [
        PHANTOMJS_EXEC,
        os.path.join(get_resource_dir('phantomjs'), 'snapshot.js'),
        file_name.replace('\\', '/'),
        file_type,
        str(__actual_delay_in_ms),
        str(pixel_ratio)
    ]
    proc = subprocess.Popen(
        proc_params, stdout=subprocess.PIPE, shell=shell_flag)
    if PY2:
        content = proc.stdout.read()
        content = content.decode('utf-8')
    else:
        content = io.TextIOWrapper(proc.stdout, encoding="utf-8").read()
    content_array = content.split(',')
    if len(content_array) != 2:
        raise OSError("No snapshot taken by phantomjs. Please make sure it is installed and available on your PATH!")
    base64_imagedata = content_array[1]
    imagedata = decode_base64(base64_imagedata.encode('utf-8'))
    if file_type in ['pdf', 'gif']:
        save_as(imagedata, output_name, file_type)
    elif file_type in ['png', 'jpeg']:
        save_as_png(imagedata, output_name)
    else:
        raise IOError(NOT_SUPPORTED_FILE_TYPE.format(file_type))


def decode_base64(data):
    """Decode base64, padding being optional.

    :param data: Base64 data as an ASCII byte string
    :returns: The decoded byte string.

    """
    missing_padding = len(data) % 4
    if missing_padding != 0:
        data += b'=' * (4 - missing_padding)
    return base64.decodestring(data)


def save_as_png(imagedata, output_name):
    with open(output_name, "wb") as f:
        f.write(imagedata)
    print('File saved in {}/{}'.format(os.getcwd(), output_name))


def save_as(imagedata, output_name, file_type):
    m = Image.open(BytesIO(imagedata))
    m.load()
    color = (255, 255, 255)
    b = Image.new('RGB', m.size, color)
    b.paste(m, mask=m.split()[3])
    b.save(output_name, file_type, quality=100)
    print('File saved in {}/{}'.format(os.getcwd(), output_name))


def get_resource_dir(folder):
    current_path = os.path.dirname(__file__)
    resource_path = os.path.join(current_path, folder)
    return resource_path
'''

**Comment by chenjiandongx** [comment]
那你等他 reiview 吧，我是 rejected 的

**Comment by leon332157** [comment]
```import io
import os
import sys
import base64
from PIL import Image
import subprocess
import webbrowser

PY2 = sys.version_info[0] == 2

if PY2:
    from StringIO import StringIO as BytesIO
else:
    from io import BytesIO

PHANTOMJS_EXEC = "phantomjs"
NOT_SUPPORTED_FILE_TYPE = "Do not support file type {}"
DEFAULT_DELAY = 1.5


def main():
    if len(sys.argv) < 2 or len(sys.argv) > 4:
        if len(sys.argv) == 2:
            webbrowser.open('https://github.com/pyecharts/pyecharts-snapshot')
            exit(0)
        else:
            print('''Usage:   snapshot {input file} {output file [png|jpeg|gif|pdf]} {delay_in_seconds}''')
            print('''         snapshot --online_help for help online.''')
            exit(-1)
    try:
        PHANTOMJS_VERSION = (subprocess.check_output([PHANTOMJS_EXEC, '--version'])).decode('utf8')
        print("\nphantomjs version: {}".format(PHANTOMJS_VERSION))
    except OSError:
        print("No phantomjs found in your PATH. Please install it!")
        sys.exit(-1)
    file_name = sys.argv[1]
    delay = DEFAULT_DELAY
    output = 'output.png'
    if len(sys.argv) >= 3:
        file_type = sys.argv[2]
        if file_type in ['pdf', 'jpeg', 'gif']:
            output = 'output.{}'.format(file_type)
        elif file_type != 'png':
            raise IOError(NOT_SUPPORTED_FILE_TYPE.format(file_type))
        if len(sys.argv) == 4:
            delay = float(sys.argv[3])  # in seconds
    make_a_snapshot(file_name, output, delay=delay)


def make_a_snapshot(file_name, output_name, delay=DEFAULT_DELAY):
    print('Generating file ...')
    file_type = output_name.split('.')[-1]
    pixel_ratio = 2
    shell_flag = False
    if sys.platform == 'win32':
        shell_flag = True
    __actual_delay_in_ms = int(delay * 1000)

    # add shell=True and it works on Windows now.
    proc_params = [
        PHANTOMJS_EXEC,
        os.path.join(get_resource_dir('phantomjs'), 'snapshot.js'),
        file_name.replace('\\', '/'),
        file_type,
        str(__actual_delay_in_ms),
        str(pixel_ratio)
    ]
    proc = subprocess.Popen(
        proc_params, stdout=subprocess.PIPE, shell=shell_flag)
    if PY2:
        content = proc.stdout.read()
        content = content.decode('utf-8')
    else:
        content = io.TextIOWrapper(proc.stdout, encoding="utf-8").read()
    content_array = content.split(',')
    if len(content_array) != 2:
        raise OSError("No snapshot taken by phantomjs. Please make sure it is installed and available on your PATH!")
    base64_imagedata = content_array[1]
    imagedata = decode_base64(base64_imagedata.encode('utf-8'))
    if file_type in ['pdf', 'gif']:
        save_as(imagedata, output_name, file_type)
    elif file_type in ['png', 'jpeg']:
        save_as_png(imagedata, output_name)
    else:
        raise IOError(NOT_SUPPORTED_FILE_TYPE.format(file_type))


def decode_base64(data):
    """Decode base64, padding being optional.

    :param data: Base64 data as an ASCII byte string
    :returns: The decoded byte string.

    """
    missing_padding = len(data) % 4
    if missing_padding != 0:
        data += b'=' * (4 - missing_padding)
    return base64.decodestring(data)


def save_as_png(imagedata, output_name):
    with open(output_name, "wb") as f:
        f.write(imagedata)
    print('File saved in {}/{}'.format(os.getcwd(), output_name))


def save_as(imagedata, output_name, file_type):
    m = Image.open(BytesIO(imagedata))
    m.load()
    color = (255, 255, 255)
    b = Image.new('RGB', m.size, color)
    b.paste(m, mask=m.split()[3])
    b.save(output_name, file_type, quality=100)
    print('File saved in {}/{}'.format(os.getcwd(), output_name))


def get_resource_dir(folder):
    current_path = os.path.dirname(__file__)
    resource_path = os.path.join(current_path, folder)
    return resource_path
```

**Comment by leon332157** [comment]
@chenjiandongx 好的

File: pyecharts_snapshot/main.py [hunk 5]
```diff
@@ -13,36 +13,42 @@
 else:
     from io import BytesIO
 
-PHANTOMJS_EXE = "phantomjs"
-NOT_SUPPORTED_FILE_TYPE = "Do not support file type %s"
+PHANTOMJS_EXEC = "phantomjs"
+NOT_SUPPORTED_FILE_TYPE = "Do not support file type {}"
```
Comment (chfw) [inline pyecharts_snapshot/main.py]: python 2.6 won't work with '{}'

File: pyecharts_snapshot/main.py [hunk 6]
```diff
@@ -13,36 +13,42 @@
...
 DEFAULT_DELAY = 1.5
 
 
 def main():
     if len(sys.argv) < 2 or len(sys.argv) > 4:
-        raise sys.exit(-1)
-
+        if len(sys.argv) == 2:
+            webbrowser.open('https://github.com/pyecharts/pyecharts-snapshot')
```
[the remark this answers is not in the document]
Comment (chfw) [reply pyecharts_snapshot/main.py]: we can show help text instead of involving a browser

File: pyecharts_snapshot/main.py (hunk 4 above)
Comment (chfw) [reply pyecharts_snapshot/main.py]: upper() 是遵循 PIL...

File: pyecharts_snapshot/main.py [hunk 7]
```diff
@@ -105,14 +110,12 @@ def save_as(imagedata, output_name, file_type):
...
+    b.save(output_name, file_type, quality=100)
+    print('File saved in {}/{}'.format(os.getcwd(), output_name))
 
 def get_resource_dir(folder):
     current_path = os.path.dirname(__file__)
     resource_path = os.path.join(current_path, folder)
     return resource_path
 
 
-def does_phantomjs_exist():
```
Comment (chfw) [inline pyecharts_snapshot/main.py]: 这个是为了更好的测试

File: pyecharts_snapshot/main.py [hunk 8]
```diff
@@ -95,8 +99,9 @@ def decode_base64(data):
 
 
 def save_as_png(imagedata, output_name):
-    with open(output_name, "wb") as g:
-        g.write(imagedata)
+    with open(output_name, "wb") as f:
+        f.write(imagedata)
+    print('File saved in {}/{}'.format(os.getcwd(),output_name))
```
Comment (chfw) [inline pyecharts_snapshot/main.py]: python 2.6 does not like this.

File: pyecharts_snapshot/main.py [hunk 9]
```diff
@@ -13,36 +13,42 @@
...
     delay = DEFAULT_DELAY
     output = 'output.png'
     if len(sys.argv) >= 3:
         file_type = sys.argv[2]
         if file_type in ['pdf', 'jpeg', 'gif']:
-            output = 'output.%s' % file_type
+            output = 'output.{}'.format(file_type)
         elif file_type != 'png':
-            raise Exception(NOT_SUPPORTED_FILE_TYPE % file_type)
+            raise IOError(NOT_SUPPORTED_FILE_TYPE.format(file_type))
```
Comment (chfw) [inline pyecharts_snapshot/main.py]: Good 👍 on explicit error

File: pyecharts_snapshot/main.py [hunk 10]
```diff
@@ -13,36 +13,42 @@
...
+            print('''Usage:   snapshot {input file} {output file [png|jpeg|gif|pdf]} {delay_in_seconds}''')
+            print('''         snapshot --online_help for help online.''')
+            exit(-1)
     try:
-        does_phantomjs_exist()
+        PHANTOMJS_VERSION = (subprocess.check_output([PHANTOMJS_EXEC, '--version'])).decode('utf8')
+        print("\nphantomjs version: {}".format(PHANTOMJS_VERSION))
     except OSError:
-        print("No phantomjs found in your path. Please install it!")
+        print("No phantomjs found in your PATH. Please install it!")
```
Comment (chfw) [inline pyecharts_snapshot/main.py]: good on specific path name

**Comment by chfw** [comment]
Please refine your PR and get them pass travis-ci first

File: pyecharts_snapshot/main.py (hunk 9 above)
Comment (chenjiandongx) [reply pyecharts_snapshot/main.py]: 看起来这个是唯一看起来 Good 的地方 😂

**Comment by chfw** [comment]
Just let you know that travis-ci also checks pep8 coding style.

**Comment by leon332157** [comment]
@chfw OK, I will figure out how to use wit 2.6!

**Comment by leon332157** [comment]
@chfw OK, I will figure out!

**Comment by leon332157** [comment]
@chfw But, is there people still using py2.6?

**Comment by leon332157** [comment]
@chfw And also that it's not necessery to use onlin help, I just thougth that it might be easier to update the document. And it is possible to show markdown in python, But require install.

**Comment by leon332157** [comment]
@chenjiandongx Costom Exception算fail？

**Comment by leon332157** [comment]
@chenjiandongx 就是enough test的那个

**Comment by chfw** [comment]
python 2.6 is mentioned in the PR title

**Comment by leon332157** [comment]
@chfw Oh! I didn't mean that 😅

**Comment by chfw** [comment]
https://www.python.org/dev/peps/pep-0008/#maximum-line-length

**Comment by chfw** [comment]
better to do local unit test first. call : `make` if you are on a linux system

**Comment by leon332157** [comment]
@chfw Seems like there're some problem in the test.sh? I can't do anything at all.

**Comment by leon332157** [comment]
@chfw It works on my mac and I'm testing on my Ubuntu

File: pyecharts_snapshot/main.py [hunk 11]
```diff
@@ -13,36 +14,41 @@
...
-        raise sys.exit(-1)
-
-    try:
-        does_phantomjs_exist()
-    except OSError:
-        print("No phantomjs found in your path. Please install it!")
-        sys.exit(-1)
-
+        if len(sys.argv) == 2:
+            webbrowser.open('https://github.com/pyecharts/pyecharts-snapshot')
```
Comment (chfw) [inline pyecharts_snapshot/main.py]: better not invite a webbrowser here. if you want to, you can put the link in the help text..

**Comment by leon332157** [comment]
@chfw Is there any difference between CI and running test?

File: tests/test_snapshot.py [hunk 12]
```diff
@@ -1,193 +0,0 @@
-import os
```
Comment (chfw) [inline tests/test_snapshot.py]: This is a no, no.

**Comment by leon332157** [comment]
Starting a new PR. @chfw @chenjiandongx

**Pull request closed without merging**